pkg/controller/migcluster/remote_watch.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 Red Hat Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migcluster
+
+import (
+	"os"
+
+	"github.com/fusor/mig-controller/pkg/controller/remotewatcher"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	migrationv1alpha1 "github.com/fusor/mig-controller/pkg/apis/migration/v1alpha1"
+	velerov1 "github.com/heptio/velero/pkg/apis/velero/v1"
+)
+
+// RemoteManagerConfig specifies config options for setting up a RemoteWatch Manager
+type RemoteManagerConfig struct {
+	// rest.Config for remote cluster to watch Velero events on
+	RemoteRestConfig *rest.Config
+	// nsname used in mapping MigCluster resources to RemoteWatchManagers
+	ParentNsName types.NamespacedName
+	// MigMigration object containing v1.Object and runtime.Object needed for remote cluster to properly forward events
+	ParentResource *migrationv1alpha1.MigCluster
+}
+
+// TODO: add support for forwarding events to multiple channels so that MigStage and
+// MigMigration controllers can also be notified of Velero events on remote clusters.
+
+// StartRemoteWatch will configure a new RemoteWatcher manager + controller to monitor Velero
+// events on a remote cluster. A GenericEvent channel will be configured to funnel events from
+// the RemoteWatcher controller to the MigCluster controller.
+func StartRemoteWatch(r *ReconcileMigCluster, config RemoteManagerConfig) error {
+	rwm := GetRemoteWatchMap()
+
+	mgr, err := manager.New(config.RemoteRestConfig, manager.Options{})
+	if err != nil {
+		log.Error(err, "[rWatch] unable to set up remote watcher controller manager")
+		os.Exit(1)
```

Comment:
I find these `Exit()` calls troubling though don't know what is common practice for controllers.  Wouldn't it be better to return the _error_  instead?

Reply:
ACK, I put this in here to get my attention with a full-on stop if the system didn't work... I agree it would be correct to return some sort of error message instead of exiting the controller.

It's not always a terrible thing for a workload in k8s to exit if it hits a fatal error since it will auto-restart assuming it's controlled under a deployment or similar, but this isn't a situation where we really want to exit, since it's possible the user provided bogus info for the remote cluster and it just needs to be fixed.